# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Pranavsanthoshnair/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [18]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 1. My Lane (or Freestyle) and Why

### My Provisional Lane: Refresh / Content Opportunity Scoring

I am choosing the **Refresh / Content Opportunity Scoring** lane provisionally.

My goal is to investigate how observable search performance, content freshness, visibility, position, and engagement signals can help prioritize pages for human review.

This lane interests me because a content team usually cannot review every page equally. The practical decision is therefore not simply whether a page is "good" or "bad", but **which pages should be reviewed first when review capacity is limited**.

The starter pipeline also showed that different approaches can rank potential opportunities differently. In the starter evaluation, the baseline rule achieved a **Precision@50 of 0.240**, while the Random Forest achieved **0.740** on the chosen starter label. This does not prove that a model will always be better, but it suggests that combining multiple observable signals may be useful for prioritization.

Based on these early results, I think this lane is worth investigating further because prioritizing pages is a practical decision problem where data may help humans focus their limited time on stronger candidates for review.

My lane is provisional and may change as I learn more about the full warehouse data and validate which signals are genuinely useful.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [19]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The Question

### Research Question

**Given observable search performance, content freshness, visibility, position, and engagement signals, which content pages should be prioritized for human review because they show evidence of decline or potential content opportunity?**

### Unit of Analysis

The unit of analysis is **one content page**.

Each observation represents a page with available signals describing its content and search performance.

### The Decision

The decision I want to support is:

> **Which pages should a content or SEO team review first when they have limited time and resources?**

A team may have hundreds or thousands of pages, so reviewing every page with the same priority is not practical.

### The Output

The intended output is a **ranked list of pages for review**.

Higher-ranked pages should represent stronger candidates for human attention based on the available evidence.

A future version of the output could include:

- a priority score;
- a suggested review category;
- reason codes explaining why the page was prioritized;
- an indication of confidence or evidence strength.

### The Action

A human reviewer could use the ranking to decide whether to:

- refresh or update content;
- investigate a possible decline;
- improve search presentation or CTR;
- expand useful content;
- protect an important page;
- monitor the page before making changes;
- consider pruning or consolidation where appropriate.

The system would support human decision-making rather than automatically changing content.

### Cost of a Wrong Recommendation

A **false positive** could cause the team to spend limited time reviewing a page that was not actually an important opportunity.

A **false negative** could cause an important declining or high-opportunity page to remain unnoticed.

Both errors have a cost because content-review capacity is limited.

For this reason, the goal is not simply to classify pages correctly. The goal is to produce a useful prioritization that helps humans decide where to spend their attention first.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [20]:
import pandas as pd
import os

repo_name = "flyrank-ml-internship"
repo_url = f"https://github.com/Pranavsanthoshnair/{repo_name}.git"
repo_path = f"/content/{repo_name}"

# Clone the repository if it doesn't already exist
if not os.path.exists(repo_path):
    print(f"Cloning {repo_url}...")
    !git clone {repo_url}
    %cd {repo_name}
else:
    print(f"Repository {repo_name} already cloned. Changing directory...")
    %cd {repo_name}

# Now, the current working directory is the repository root
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print(f"Total pages: {len(df):,}")
print(f"Total columns: {len(df.columns)}")

declining = (df["trend_direction"].str.lower() == "down").sum()
declining_pct = declining / len(df) * 100

print(f"Declining pages: {declining:,}")
print(f"Declining percentage: {declining_pct:.2f}%")

Repository flyrank-ml-internship already cloned. Changing directory...
[Errno 2] No such file or directory: 'flyrank-ml-internship'
/content/flyrank-ml-internship
Total pages: 30,000
Total columns: 44
Declining pages: 16,262
Declining percentage: 54.21%


In [21]:
visible_declining = df[
    (df["trend_direction"].str.lower() == "down") &
    (df["impressions_90d"] >= 500)
]

print(f"Declining pages with >= 500 impressions: {len(visible_declining):,}")

Declining pages with >= 500 impressions: 9,961


## 3. Quick Look at the Data

I used the starter dataset to check whether the **Refresh / Content Opportunity Scoring** lane appears worth investigating.

### Dataset Overview

The starter dataset contains **30,000 pages** and **44 columns**.

This provides a reasonably large set of observations for exploring differences in page performance and identifying patterns that may help prioritize pages for human review.

### Declining Pages

The dataset contains **16,262 pages labelled as declining**, representing **54.21% of all pages** in the starter dataset.

This shows that declining performance is not limited to just a small number of pages. Since a team cannot realistically investigate every declining page with equal attention, prioritization becomes an important practical problem.

### Declining Pages With Meaningful Visibility

Among the declining pages, **9,961 pages have at least 500 impressions**.

These pages are particularly relevant to my chosen lane because they combine evidence of decline with meaningful search visibility.

A declining page with very little visibility may have a different priority from a declining page that is still receiving substantial impressions. This suggests that prioritization may need to consider multiple signals rather than simply treating every declining page equally.

### Why These Numbers Support My Lane

The starter data shows substantial variation in page performance and visibility. With **16,262 declining pages** and **9,961 of those still receiving at least 500 impressions**, there is a clear practical need to decide which pages deserve attention first.

These numbers do not prove that any particular feature causes a page to decline. However, they support investigating whether combinations of observable signals can help create a more useful ranking for human review.

This is why I believe the **Refresh / Content Opportunity Scoring** lane is worth investigating over the next several weeks.


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [22]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 4. Careful Words: What I Can and Can't Claim

### What I Can Claim

I can describe patterns in this dataset as:

- observed;
- measured;
- associated;
- correlated;
- directional.

I can compare different ranking or prediction approaches using clearly defined evaluation methods.

I can investigate whether combinations of available signals appear useful for prioritizing pages for human review.

### What I Cannot Claim

I cannot claim that:

- I have discovered how Google's ranking algorithm works;
- correlation proves that one feature causes another outcome;
- a page recommended for review will definitely recover after changes are made;
- a high model score guarantees that a page should be changed;
- a model can replace human judgment.

### How I Will Frame This Project

This project is intended to be a **decision-support system**.

Its purpose is to help prioritize pages for human review using the available evidence.

A recommendation means:

> **This page may deserve attention earlier than other pages based on the available signals.**

It does **not** mean:

> **Changing this page is guaranteed to improve its performance.**

Any future model will also need careful validation and protection against data leakage. Features used for prediction should be information that would actually be available at the time the recommendation is made.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.